### DISCUSSION TITLE
Authentication Servers

### DISCUSSION THREAD

**Discussion by 1WHISKY** [stoatchat/discussions#103] [Feature Suggestions]
### Issue 
If you join five different Revolt (self hosted) instances you have to create five different accounts for each instance.  
This can get annoying because you have to create a account even if you are planning to join that instance temporarily.

Also you have to trust the hoster that he did not tamper with the service in order to sniff your entered credentials.

### Possible Solution
Create a publicly hosted authentication server where people can register.
Additionally you can host your own authentication server.

Once you join a self hosted instace you can choose a authentication server using a drop-down menu.
If you choose to log in using the public authentication server you will be redirected to something like auth.revolt.chat and once you logged in you will be returned to the self hosted instance.

The available/allowed authentication servers should be configurable (e.g. .env).
Maybe other login providers (oauth2) could be included so you can log in with steam, google etc..

### Problems
There are probably many issues with this suggestion and my suggested solution is probably not the best.
What happens if two people join a self hosted instace with the same name from two different authentication servers?

Feel free to comment how you would implement such an authentication system

**Comment by insertish** [comment]
> What happens if two people join a self hosted instace with the same name from two different authentication servers?

Ideally I would do it so that each instance can only connect to one other authentication server.

**Reply by Eragonfr** [reply]
An other method is to federate the autentication servers. So only one person can claim a username

**Reply by surfaceflinger** [reply]
> So only one person can claim a username

IMO it would be better if user had some kind of immutable "friend code" derived from private key stored on an authentication server, and they could set any display name they want. This would also prevent the "OG" cancer.

**Comment by mkg20001** [comment]
Another thing to consider is OAuth2 (Sign-In via Google, etc)

Self-hosted servers could, maybe even by default, offer "Sign-In with revolt.chat Account"

**Reply by oleggtro** [reply]
We could support both. I really like the idea of federating servers. That would essentially mean that someone could sign in to a server with any account. Just append the domain to the username and we could request the required info from the users home server. Somewhat like matrix or mastodon usernames

**Reply by Eragonfr** [reply]
That's exactly how I see it. It's a little constraint for the user (you need to remember your instance domain name) but in exchange you can use one account and one instance to interact with any other one.

**Reply by mkg20001** [reply]
OpenID allows doing more-or-less what the federated use-case would entail (sign-in with different domain, sharing profile information). This way it would be standards-compatible. Users could then pick from "sign-in via google, facebook, etc" or "sign in with different revolt instance: [domain]"

**Reply by oleggtro** [reply]
I think we shouldn't support sign in via google cuz that would require a central instance handling all the oauth stuff. For example when user A (registered on instance 1) wants to join server xyz hosted on instance 2 he would have to put in his username, then get a oauth popup of his homeinstance to authorize and then he could join. If we would want to support Google OAuth there would be a central dictionary of all accounts with connected google accounts to match and get the oauth creds from the users homeserver. Its pretty useless as that would introduce a single point of failure and suggests the users a false privacy

**Reply by mkg20001** [reply]
OAuth is already "central" as in the sense that the server behind your-revolt.tld handles confirming whether the user exists and also giving the other server the profile. Everything would be as-usual, just whenever signing in to your-revolt.tld you could either do username/pw or additionally google.

Edit: Sign-in via google is just an example, any server could add anything there that supports oauth.

**Reply by oleggtro** [reply]
Yea that would be possible, but i would recommend limiting it for privacy reasons to accounts registered on the instance you are signing into.

**Reply by mkg20001** [reply]
I mean the following setup 

`revolt instance <-> other revolt instance <-> external oauth`

Meanning you'd have "federation-oauth" (you sign in as user@instance2 on instance1) and then external oauth sign-in (optional) with an external party (github, gitlab, whatever)

There's no breach of privacy if you don't press "sign-in with whatever" on that server.

Matrix can also handle oauth-sign-in just fine. https://matrix.org/docs/guides/sso-for-client-developers

**Reply by oleggtro** [reply]
won't be implemented anyway  https://github.com/revoltchat/revolt/discussions/125#discussioncomment-1341921

**Reply by mkg20001** [reply]
Sad to see. Understandable why. OAuth sign-in would still be a nice addition, saves the hassle of an extra password.

**Reply by oleggtro** [reply]
Just use a password manager lmao

**Reply by mkg20001** [reply]
Ey get your ego somewhere else. It's a useful feature for some people.

**Reply by oleggtro** [reply]
ok, but we're really off topic now

**Reply by tallship** [reply]
# OpenID/OAuth2 Plugin capability

### Service can be easily enabled based on administrator preferences to support other AUTH/SSO methods

> Another thing to consider is OAuth2 (Sign-In via Google, etc)

> Self-hosted servers could, maybe even by default, offer "Sign-In with revolt.chat Account"

## Basics of enabling the capabilities like many other platforms services do:

This is a good idea. I don't think it is necessarily a good idea to be worrying about it  at this **particular** time however, but many sites afford the userbase the opportunity to either create a new account or **connect** their account to a third party authenticator service. platforms like Discourse, Dev.to, NextCloud, and others enable this (in addition to many that also offer Google, Faceplant, GitHub, etc.).

This is typically enabled by installing the capability  via the native plugin architecture of platform, so that it would be a decision on the part of each Revolt instances administrators.

Many of us already offer these capabilities for our communities, especially those of us who operate many different types of platforms, providing an SSO solution for  those users who choose to enable it for themselves on their respective accounts; and many of us also operate the OpenID/OAuth2 provider services to facilitate this. One free solution that I personally like is:

https://github.com/DuendeSoftware/IdentityServer (Easy Peasy to deploy)

Again, this wouldn't be built into the core of Revolt, but rather, as an optional service/plugin to enable depending on the needs/preferences of the administrators of other self-hosted Revolt instances. The Revolt server admins could also choose from amongst other various providers (besides their own OpenID/OAuth2 provider service) like GitHub, and the evil monolithic silos like Google, Twitch, and Faceplant.

The flow would work this way:

* User signs up by creating a new email based account on the Revolt server, then later, choses to connect the SSO Auth service to their account - thereafter, they can login using either method (i.e., using OpenID/OAUTH2 or via email/username).
* User signs up with the external auth provider where they already have an identity (i.e, using their OpenID/OAUTH2 provider account).

Other Revolt and different kinds of platform instances independently decide whether they want to offer these additional types of authentication methods at their own prerogative.

I'm interested in hearing other's thoughts on the matter, and will eventually probably open an issue  directly related once Revolt matures a bit - in the meantime, the  development cycle of the project is seemingly too aggressive and tending to more immediate matters/issues on the projects roadmap to detract from such a tertiary plugin integration.

⛵ 

.

**Reply by erlend-sh** [reply]
> Self-hosted servers could, maybe even by default, offer "Sign-In with revolt.chat Account"

For this I highly recommend https://github.com/sebadob/rauthy

**Comment by oleggtro** [comment]
maybe rename this to `Server Federation` ? @1WHISKY

**Comment by walmir** [comment]
Hello,
Is there an update on this? 
I need this on a self-hosted instance. The goal is to create a single account on an auth-server for multiple apps including revolt.chat.
If this hasn't been done, or planned yet, I'm considering implementing it myself. 
I'm happy to discuss other requirements and sharing results :)

**Comment by Fifthdread** [comment]
I also need Authentication Server support for my self-hosted instance. Specifically I want to use Authentik. I hope it gets implemented at some point.

**Comment by Zipdox2** [comment]
I second using the official Revolt server as an Oauth provider for self-hosted instances.

**Comment by secondtruth** [comment]
That is a good idea for now, but in the long term, I think Federation would be the best solution to keep centralization low. There's also the emerging [Polyproto protocol](https://docs.polyphony.chat/Overviews/core/), which is modeled after Discord and could be implemented instead of ActivityPub or Matrix.

See @erlend-sh's [reply to another post](https://github.com/orgs/revoltchat/discussions/125#discussioncomment-9327808) for more details about this and federation in general.

**Reply by Zipdox2** [reply]
I don't think Polyproto passes the grandma test. I've been brainstorming an alternative system.

**Reply by bitfl0wer** [reply]
hmm

**Reply by Zipdox2** [reply]
> hmm

I decided that for a federated chat app to succeed it needs to be purpose-built application and not some generic federation protocol detached from any specific implementation. I have been brainstorming an architecture that I may turn into an actual project. If you are interested in contributing let me know.

**Reply by tallship** [reply]
Perhaps as a plugin, which 3rd party providers could leverage for other, additional networks?  Something that implements this methodology for example, is Hubzilla - even for ActivityPub, the individual user must activate the plugin to enable their own account to engage with that network.

**Reply by bitfl0wer** [reply]
> > hmm
> 
> 
> 
> I decided that for a federated chat app to succeed it needs to be purpose-built application and not some generic federation protocol detached from any specific implementation. I have been brainstorming an architecture that I may turn into an actual project. If you are interested in contributing let me know.

I am just interested in why you think this way about p2, and I'd like to hear your reasoning behind it, is all. polyproto is purpose-built to make an app which does not need to do much explaining at all to people unfamiliar with the concept of federation, so this feedback makes me curious. Since this is a thread about Stoat, do feel free to write an email about it to flori@polyphony.chat, or a message to our forum at https://polyphony.zulipchat.com/

**Reply by Zipdox2** [reply]
> I am just interested in why you think this way about p2, and I'd like to hear your reasoning behind it, is all.

What is p2 in this case?

**Reply by anotherdoesnm** [reply]
> > I am just interested in why you think this way about p2, and I'd like to hear your reasoning behind it, is all.
> 
> What is p2 in this case?

PolyProto (repeated P are changed with 2)
like I2P - Invisible Internet Project

**Comment by terion-labs** [comment]
SSO is a must for such product

**Comment by Obnyr** [comment]
I disagree. Let's say one day the authentification server closes. What do you do ? How do you get your things back ? Having multiple accounts is the right way to go, so even if one host fails, or if hosts can't communicate between them anymore, you can still access your data everywhere. Centralisation should be avoided.

However, it is true that creating an account for every host is exhausting. A tool to automate account generation on a new host is not a bad idea. Like you could link a vault to your session, and the newly automatedly created account would then be added to the vault like KeePassXC or Bitwarden. That way, you don't have any hassle creating a new account, a new password etc, but you keep access to them.

**Comment by fhibler** [comment]
I can only second that centralized authentication (Oauth/SSO) would be highly desireable. I would also not mix this us with server federation, but pure 'account and permission management'.

**Comment by konstantinschuette** [comment]
https://github.com/stoatchat/stoatchat/discussions/507#discussioncomment-15710135
I found this. At the beginning of the year OIDC was almost finished. Not sure how things are now
